src/assignment-5/CollateralizedVault.sol

```diff
@@ -0,0 +1,146 @@
+        priceFeed = AggregatorV3Interface(priceFeed_);
+    }
+
+    /// @notice Deposits additional collateral into the vault
+    /// @param collateralAmount amount of `collateral` token to deposit
+    function deposit(uint256 collateralAmount) public {
+        depositedCollateral[msg.sender] += collateralAmount;
+
+        collateral.transferFrom(msg.sender, address(this), collateralAmount);
+    }
+
+    /// @notice Borrows `underlying` token from the Vault
+    ///     Only allowed to borrow up to the value of the collateral
+    /// @param amount The amount of `underlying` token to borrow
+    function borrow(uint256 amount) public {
+        uint256 maxDebt = getMaxAllowedDebt(depositedCollateral[msg.sender]);
+        if (debt[msg.sender] + amount > maxDebt) {
+            revert NotEnoughCollateral();
+        }
+
+        debt[msg.sender] += amount;
+
+        underlying.transfer(msg.sender, amount);
+    }
+
+    /// @notice Pays back an open debt
+    /// @param amount The amount of `underlying` token to pay back
+    function repayDebt(uint256 amount) public {
+        debt[msg.sender] -= amount;
+
+        underlying.transferFrom(msg.sender, address(this), amount);
+    }
+
+    /// @notice Withdraws part of the collateral
+    ///     Only allowed to withdraw as long as the collateral left is higher in value than the debt
+    /// @param collateralAmount The amount of `collateral` token to withdraw
+    function withdraw(uint256 collateralAmount) public {
+        uint256 requiredCollateral = getRequiredCollateral(debt[msg.sender]);
+
+        if (depositedCollateral[msg.sender] - collateralAmount < requiredCollateral) {
+            revert TooMuchDebt();
+        }
+
+        depositedCollateral[msg.sender] -= collateralAmount;
+
+        collateral.transfer(msg.sender, collateralAmount);
+    }
+
+    /// @notice Admin: liquidate a user debt if the collateral value falls below the debt
+    /// @param user The user to liquidate
+    /// @dev Only admin is allowed to liquidate
+    function liquidateUser(address user) onlyOwner public {
+        uint256 requiredCollateral = getRequiredCollateral(debt[user]);
+        if (depositedCollateral[user] >= requiredCollateral) {
+            revert UserDebtIsSufficientlyCollateralized();
+        }
+
+        delete debt[user];
+        delete depositedCollateral[user];
+    }
+
+    /// @notice Returns the maximum allowed debt for the given `collateralAmount`
+    function getMaxAllowedDebt(uint256 collateralAmount) public view returns (uint256 underlyingAmount) {
+        underlyingAmount = (collateralAmount * 10**priceFeed.decimals()) / getPrice();
+        underlyingAmount = scaleInteger(underlyingAmount, collateral.decimals(), underlying.decimals());
+    }
+
+    /// @notice Returns the required amount of collateral in order to borrow `borrowAmount`
+    function getRequiredCollateral(uint256 borrowAmount) public view returns (uint256 requiredCollateral) {
```

Comment:
Each `.decimals()` call is 2100 gas. How could you get rid of them?

Reply:
@alcueca I could save them in the contract storage in the constructor.
Is that what you had in mind?

Reply:
That's probably best 👍

Reply:
done: https://github.com/davidbrai/yield-mentorship/pull/6/commits/993c3285bd6873240f8d01e24afd5f922577f68a